### ISSUE TITLE
Shared subscriber fail over issue

### ISSUE THREAD

**Issue by shaloba** [erlio/vernemq#751]
### Environment

- VerneMQ Version: 1.4.0
- OS: Ubuntu

Hi, 
I with to handle subscriber failure and message recovery (if one subscriber goes down its messages will be moved to the other subscribers of the topic)
I'm using multiple subscribers on single topic ($share/group/topic1)

**I made the following steps:**

1. set "persistent_client_expiration" to 1d and restart vernemq service
2. run 8 subscribers to "topic1" with different client id and clean_session set to False
3. publish 1000 messages to "topic1" 
4. all the subscribers shared the messages equally 
5. stop the 8 subscribers
6. publish 1000 messages to "topic1"
7. verify the messages pending for each client using "vmq-admin session show --topic  --client_id --is_online --queue_size --online_messages --offline_messages"
8. start 7 of the 8 subscribers- all messages pulled except client-8 (still stopped)
9. check the broker status day after

### Expected behavior
1. client-8 disappear from the broker clients  
2. client-8 messages should shared between the other clients

### Actual behaviour

1. client-8 disappear from the broker clients  
2. messages didn't pulled by the other 7 clients

Am  I doing something wrong ?!? :/

**Comment by ioolkos** [comment]
Hi @shaloba thanks for asking!

You're not doing anything wrong, and everything works as expected... but yes, this is still unfortunate, as the client subscription expires.

In the MQTT v5 spec for shared subscription it is explicitly forbidden to re-route those messages:

>If the Client's Session terminates before the Client reconnects, the Server MUST NOT send the Application Message to any other subscribed Client [MQTT-4.8.2-5].

That said, as long as you have shared subscribers remaining online, they should get the messages directly.
The broker only fills up offline queues, when there are nothing but offline queues.

**Comment by shaloba** [comment]
Thanks for your answer!
It's very unfortunate to hear that. does Verne planning to create a patch to this issue or its MQTT protocol restrictions and you will not change its flow ?

**Comment by dergraf** [comment]
I guess if the demand for such a feature is big enough or someone considers to sponsor such a feature we would have to think about it. 
As shared subscription is actually a MQTT5 feature (that we have implemented/backported for MQTT4) I think it might makes sense to wait with non-standard features until we have 'officially' released MQTT5 (merged in master branch) and have the first round of bugs fixed for the standard features.

But, this is just me, what do you think @larshesel

**Comment by larshesel** [comment]
@dergraf I agree with that approach. Let's think about these concerns after MQTTv5 has been merged to the master branch. I do could envision some feature like the above. It might be more general though were we can do some action when a session is expired (one of them could be to republish messages).

**Comment by francois-travais** [comment]
@larshesel @dergraf I'm also interested about this feature, do you have any update?

**Comment by larshesel** [comment]
There's no work in progress for this feature at the moment.
